src/nbl/video/CVKSwapchain.cpp

```diff
@@ -58,39 +58,40 @@ CVKSwapchain::CVKSwapchain(SCreationParams&& params, CVKLogicalDevice* dev) : IS
         params.mipLevels = 1u;
         params.samples = CVulkanImage::ESCF_1_BIT;
         params.type = CVulkanImage::ET_2D;
-
+        
         // TODO might want to change this to dev->createImage()
```

Comment:
actually, `createImage` is only for making a new `VkImage` 

@Crisspl should `CVulkanImage` hold an optional backpointer to a swapchain?

Reply:
`CVulkanImage::~CVulkanImage` needs to know if it should call `vkDestroyImage`.

1. If this image is part of a swapchain and we haven't created the image ourselves (for example, by extending `VkSwapchainCreateInfo` structure) then `vkDestroyImage` should not be called.
 
2. If this image is part of a swapchain and we created it ourselves, then `vkDestroyImage` should be called.

3. If this image is not part of a swapchain, then `vkDestroyImage` should be called.

This means that merely keeping a backpointer to the swapchain couldn't help us decide if we should destroy the image.
Basically, if `vkCreateImage` was called then `vkDestroyImage` should also be called, so it'd make sense to put these calls in constructor-desctructor pair of `CVulkanImage`. But there are two problems with it:

1. https://github.com/Devsh-Graphics-Programming/Nabla/pull/160#discussion_r676511378
2. `ISwapchain` has storage for `IGPUImage` which may or may not be required.

Reply:
probably should just keep a member flag to say something like:
```C
CVulkanImage::~CVulkanImage()
{
    if (m_imageShouldBeDestroyed)
        vkDestroyImage();
}
```

Reply:
@achalpandeyy create a `CVulkanImage` and `CVulkanForeignImage : public CVulkanImage` , because if we start importing images by their `vkImage` handle from Unreal, Unity, Qt6 we also can't delete them and seems like `ForeignImage` is the best solution for this and the swapchain

Just create the correct type and do this
```cpp
// m_ownerKeepalive is core::smart_refctd_ptr<IReferenceCounted>
CVulkanForeignImage::~CVulkanForeignImage()
{
   m_ownerKeepalive = nullptr; // for Swapchain images, its a pointer to swapchain, for some other foreigner optional
   m_image = VK_INVALID_HANDLE;
}

CVulkanImage::~CVulkanImage()
{
   if (m_image!=VK_INVALID_HANDLE)
      vkDestroyImage(m_image);
}
```

# Understanding Refcounting of Vulkan objects

btw if you pass your own images to CreateSwapchain, then swapchain needs to refcount the images (because its a user) as the swapchain's existence is dependent on the existence of the images (if you delete the images and you keep using the  swapchain you have a problem)

If your CreateSwapchain makes images that you then need to retrieve, it means that the images need to refcount the swapchain because their existence is dependent on the existence of the swapchian (if you delete the swapchain and keep using the images its a problem)

Reply:
does your Foreign Image keep a refcounted backpointer to swapchain?

Reply:
theres also a circular dep here

Reply:
To be continued here https://github.com/Devsh-Graphics-Programming/Nabla/issues/246